### PR TITLE
Honor explicit Durable HTTP request timeouts

### LINKED ISSUE #1724: DurableHttpRequest.Timeout throws System.TimeoutException after wrong time.
### Description
I'm not sure if it is really a bug or am I doing something wrong.
I'm calling `await context.CallHttpAsync(request)` with `var request = new DurableHttpRequest(HttpMethod.Post, new Uri(_configuration.TaggingUrl), headers, body, timeout: TimeSpan.FromMinutes(30));` to api that takes long time. But it throws timeout exception much faster than that. It happens during local debuging. Haven't tested on Azure. Bellow are logs
```
[2021-03-12T14:11:56.091Z] Executing 'Tagging_HttpStart' (Reason='This function was programmatically called via the host APIs.', Id=e2cbe143-fd5d-4f5d-851c-4b010f95d2e4)
[2021-03-12T14:11:56.300Z] Started orchestration with ID = '281a0a0fab754831bbc843663799bbf8'.
[2021-03-12T14:11:56.329Z] Executed 'Tagging_HttpStart' (Succeeded, Id=e2cbe143-fd5d-4f5d-851c-4b010f95d2e4, Duration=269ms)
[2021-03-12T14:12:00.967Z] Host lock lease acquired by instance ID '00000000000000000000000022A5C165'.
[2021-03-12T14:12:01.687Z] Executing 'Tagging' (Reason='(null)', Id=ab51098d-5b58-4f65-bbfb-77997cc1f35d)
[2021-03-12T14:12:01.707Z] Executed 'Tagging' (Succeeded, Id=ab51098d-5b58-4f65-bbfb-77997cc1f35d, Duration=26ms)
[2021-03-12T14:12:01.773Z] Executing 'Tagging_GetItems' (Reason='(null)', Id=2f7dc096-af44-408e-8c70-bb0504b8c2ff)
[2021-03-12T14:12:07.635Z] Executed 'Tagging_GetItems' (Succeeded, Id=2f7dc096-af44-408e-8c70-bb0504b8c2ff, Duration=5874ms)
[2021-03-12T14:12:07.911Z] Executing 'Tagging' (Reason='(null)', Id=37c7d063-786b-40e1-8adf-3eed92732204)
[2021-03-12T14:12:07.984Z] Executed 'Tagging' (Succeeded, Id=37c7d063-786b-40e1-8adf-3eed92732204, Duration=72ms)
[2021-03-12T14:16:08.169Z] Executing 'Tagging' (Reason='(null)', Id=45362154-60ec-4e15-a12d-474f739ffac0)
[2021-03-12T14:16:08.313Z] 281a0a0fab754831bbc843663799bbf8: Function 'Tagging (Orchestrator)' failed with an error. Reason: System.TimeoutException: The operation was canceled. Reached user specified timeout: 00:30:00.
[2021-03-12T14:16:08.315Z]    at Microsoft.Azure.WebJobs.Extensions.DurableTask.DurableOrchestrationContext.CallDurableTaskFunctionAsync[TResult](String functionName, FunctionType functionType, Boolean oneWay, String instanceId, String operation, RetryOptions retryOptions, Object input, Nullable`1 scheduledTimeUtc) in D:\a\r1\a\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\ContextImplementations\DurableOrchestrationContext.cs:line 698
[2021-03-12T14:16:08.316Z]    at Microsoft.Azure.WebJobs.Extensions.DurableTask.DurableOrchestrationContext.ScheduleDurableHttpActivityAsync(DurableHttpRequest req) in D:\a\r1\a\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\ContextImplementations\DurableOrchestrationContext.cs:line 283
[2021-03-12T14:16:08.316Z]    at Microsoft.Azure.WebJobs.Extensions.DurableTask.DurableOrchestrationContext.Microsoft.Azure.WebJobs.Extensions.DurableTask.IDurableOrchestrationContext.CallHttpAsync(DurableHttpRequest req) in D:\a\r1\a\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\ContextImplementations\DurableOrchestrationContext.cs:line 246
[2021-03-12T14:16:08.317Z]    at Iine.AzureFunction.Tagging.Functions.TaggingFunctions.RunOrchestrator(IDurableOrchestrationContext context) in C:\Users\piotr\Documents\Repos\Source\iine-datafactory\AzureFunctions\Iine.AzureFunction.Tagging\Functions\TaggingFunctions.cs:line 44
[2021-03-12T14:16:08.318Z]    at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2.InvokeAsync(Object instance, Object[] arguments) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionInvoker.cs:line 52
[2021-03-12T14:16:08.318Z]    at Microsoft.Azure.WebJobs.Extensions.DurableTask.TaskOrchestrationShim.InvokeUserCodeAndHandleResults(RegisteredFunctionInfo orchestratorInfo, OrchestrationContext innerContext) in D:\a\r1\a\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\Listener\TaskOrchestrationShim.cs:line 150. IsReplay: False. State: Failed. HubName: TestHubName. AppName: . SlotName: . ExtensionVersion: 2.4.1. SequenceNumber: 10. TaskEventId: -1
[2021-03-12T14:16:08.329Z] Executed 'Tagging' (Failed, Id=45362154-60ec-4e15-a12d-474f739ffac0, Duration=159ms)
[2021-03-12T14:16:08.331Z] System.Private.CoreLib: Exception while executing function: Tagging. System.Private.CoreLib: Orchestrator function 'Tagging' failed: The operation was canceled. Reached user specified timeout: 00:30:00.
```

### Expected behavior
While setting timeout to TimeSpan.FromMinutes(30) - System.TimeoutException is thrown after 30 minutes

### Actual behavior
While setting timeout to TimeSpan.FromMinutes(30) - System.TimeoutException is thrown after just couple minutes.

### Relevant source code snippets
```csharp
var request = new DurableHttpRequest(
    HttpMethod.Post, new Uri(_configuration.TaggingUrl), headers, body,
    timeout: TimeSpan.FromMinutes(30));

var response = await context.CallHttpAsync(request);
```

### Known workarounds
None

### App Details

- **Durable Functions extension version**: v2.4.1
- **Azure Functions runtime version**: In host.json it is set to `"version": "2.0"` Other versions bellow.
- **Programming language used**: C#

### Screenshots
image

Zrzut ekranu 2021-03-12 151643

### PR DESCRIPTION
**Pull request by YunchuWang**
# Summary
## What changed?
- Initialize the cached Durable HTTP client with an infinite client timeout, then enforce the explicit request timeout (or the existing 240-second default) with a per-request cancellation source.
- Report an explicit timeout only when that request's own cancellation source expires, preserving the original cancellation exception as the inner cause. Do not mutate the shared client on each send.
- Add focused regression tests for factory behavior, timeout/cancellation handling, concurrency, and boundary values.

## Why is this change needed?
- A request configured for more than four minutes currently fails at the shared client's 240-second timeout while incorrectly claiming the requested timeout was reached.
- The default remains 240 seconds per HTTP exchange. No public configuration option, retry-policy change, or redesign of the overall HTTP 202 polling budget is included.

## Issues / work items
- Closes #1724
- Related: N/A

---

# Project checklist
- [x] Documentation changes are not required
  - [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
- [ ] Release notes are not required for the next release
  - [ ] Otherwise: Notes added to `release_notes.md`
- [ ] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [x] No extra work is required to be leveraged by OutOfProc SDKs
  - [ ] Otherwise: Work tracked here: #issue_or_pr_in_each_sdk
- [x] No change to the version of the `WebJobs.Extensions.DurableTask` package
  - [ ] Otherwise: Major/minor updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`
- [x] No EventIds were added to `EventSource` logs
  - [ ] Otherwise: Ensure EventIds are within the supported range in the existing Windows infrastructure (validate via deployed telemetry).
- [ ] This change should be added to the `v2.x` branch
  - [ ] Otherwise: This change applies exclusively to `WebJobs.Extensions.DurableTask` v3.x and will be retained only in the `dev` and `main` branches
- [ ] Breaking change?
  - [ ] If yes:
    - Impact: Intended behavior correction: longer explicit timeouts are honored, and unrelated cancellation is no longer reported as an explicit timeout. Default cancellation retains its exception family, but no longer reports that the client's own timeout elapsed.
    - Migration guidance: No new API or host configuration is required.
---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [ ] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [x] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot CLI.
- AI-assisted areas/files: `DurableHttpClientFactory.cs`, `TaskHttpActivityShim.cs`, and `TaskHttpActivityShimTests.cs`.
- What you changed after AI output: No code changes after the approved independent read-only review, which reported no findings. Verification details are below.

AI verification (required if AI was used):
- [ ] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing
## Automated tests
- Result: Passed.
- Permanent fast selector rerun before commit: 22 passed on .NET 8 and 22 passed on .NET 10 (12 new cases plus 10 existing request tests).
- Existing timeout, retry, and asynchronous polling selectors: 8 passed on each target framework.
- The new suite failed against the previous implementation for the expected reasons before the fix. Permanent tests use the actual factory and activity; no four-minute waits were added to routine CI.

## Manual validation (only if runtime/behavior changed)
- Environment (OS, .NET version, components): **Local Windows runs only; no Azure deployment.** .NET SDK 10.0.303; .NET 8.0.30 and .NET 10.0.11; source host extension 3.15.0 and isolated worker extension 1.19.0; Functions Core Tools 4.10.0; Functions runtime 4.1048.200.26180; Azurite 3.37.0.
- Steps + observed results:
  1. In-process WebJobs JobHost with the DTFx local emulator and a cancellation-aware handler: a 600-second request timeout allowed a 250-second response to complete with HTTP 200 (250.015 seconds measured).
  2. A real isolated Functions host and separate worker, using Azurite and loopback HTTP: the same 600-second timeout allowed a 250-second response to complete with HTTP 200 (250.005 seconds measured). The loaded host and worker assemblies matched the source builds.
  3. Concurrent controls retained the null/default cutoff near 240 seconds in both runtimes (240.003 / 240.010 seconds), and explicit one-second timeouts still expired (1.012 / 0.999 seconds). The original cancellation chain was retained for explicit timeouts.
- Evidence (optional): Before the fix, the same 600-second request timed out near 240 seconds. After the fix, both local runtime paths completed beyond that cutoff.

---

# Notes for reviewers
- Release-note and backport decisions are left unchecked for maintainer triage; this PR targets `dev` only.
- The human-author understanding checkbox remains for author confirmation; the technical verification boxes reflect the agent's checks and independent review.
- Existing isolated-worker exception-transport details and overall HTTP 202 polling behavior are intentionally unchanged.

### CODE REVIEW

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1 [hunk 1]
```diff
@@ -20,138 +25,180 @@ $funcLogId = [guid]::NewGuid().ToString("N")
...
+    param($Process)
+
+    # The npm launcher owns the native host and worker. Capture descendants before
+    # stopping it so they cannot become orphaned and hold redirected output open.
+    $allProcesses = @(Get-Process)
+    $ownedProcesses = @($Process)
+    for ($i = 0; $i -lt $ownedProcesses.Count; $i++) {
+        $parentId = $ownedProcesses[$i].Id
+        $ownedProcesses += @($allProcesses | Where-Object { $_.Parent.Id -eq $parentId })
+    }
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1]: Addressed in f1dd88a794155e1004cc79291d3f81b67cea9bd8 by adding `#Requires -Version 7.0`. PowerShell 7 provides the standard `Process.Parent` extended property on Windows and Linux, and CI already runs this script with `pwsh`.

Windows PowerShell 5.1 is now rejected before any script body or host startup executes, rather than potentially missing descendants during cleanup. The existing owned-process-tree cleanup is unchanged.

File: test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1
Comment (bachuv) [inline test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1:1]: Why do we need to make changes to run-smoke-tests.ps1?

Comment (YunchuWang) [reply test/SmokeTests/OOProcSmokeTests/DotNetIsolated/run-smoke-tests.ps1:1]: removed

**Review by bachuv — APPROVED** [review]
